src/pow.cpp

```diff
@@ -231,13 +344,115 @@ bool CheckNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHeader& bl
     return ((min <= target) && (target <= max));
 }
 
-bool CheckProofOfWork(uint256 hash, unsigned int nBits, const Consensus::Params& params)
+bool CheckNextWorkRequiredAux(const CBlockIndex* pindexLast, const CBlockHeader& block, const Consensus::Params &params)
 {
-    bool fNegative;
-    bool fOverflow;
+    // There's nothing to check before activation of merged mining.
+    if (block.m_aux_pow.IsNull()) {
```

Comment:
Verify (will delete if I manage to verify this myself): once aux pow is required, passing a block with a null aux pow should not auto-true (i.e. condition is checked everywhere before entering this function).

Reply:
Verification: this is called from a single place, `main.cpp:3837` (at the point of writing).

https://github.com/tradecraftio/tradecraft/blob/f16e7f2fa7a0115acc75f4ec80432a9bef7b5a57/src/main.cpp#L3836-L3845

It includes a pre-check for non-nullness. This code later checks that aux pow is non-null if merged mining is enabled.

Suggestion: switch to an assert here instead. This seems like a serious error that should never happen.

Reply:
Is there a reason to prefer an assert? An assert kills the node, which seems a rather dangerous thing to do based on the content of a block, which is data received by a peer.

Reply:
It's not just based on the content of a block, it's also based on the logic of the code. If the code here for some reason NOP'd line 3836, the pow check will silently accept a missing aux pow. The consequences of this happening over an extended period of time, vs. nodes crashing immediately upon it happening, raising flags and complaints.

While there are checks later on that ensure the aux pow isn't null (iirc), anything beyond this point would have the invalid assumption that the aux pow was indeed valid.

If you are concerned about an assertion, I would at least add a logprint of a warning, since the if case in `pow.cpp` should never, ever be called, at least not with current code base.